test/foundry/Regular_Tests/CounterFactuals.t.sol

```diff
@@ -0,0 +1,250 @@
+     *      were staked. 
+     */
+    function testStakeAndWithdraw() public {
+        uint256 amount = 100 ether;
+        aurora.approve(address(jetStaking), amount);
+        jetStaking.stake(amount);
+        vm.expectRevert("NOT_ENOUGH_STAKE_BALANCE");
+        jetStaking.unstake(amount + 1);
+    }
+
+    /**
+     * @dev Stakes 10**19 aurora tokens for user1 and user2. waits 10 weeks, and 
+     *      unstakes all tokens for both users. This test checks that both users
+     *      will end up with the same amount of aurora (ie. withdrawal order has
+     *      no effect).   
+     */
+    function testWithdrawalOrder() public {
+        stakeFor(user1, 10 ether);
+        stakeFor(user2, 10 ether);
+        vm.warp(block.timestamp + 10 weeks);
+        vm.startPrank(user1);
+        jetStaking.unstakeAll();
+        vm.stopPrank();
+        vm.startPrank(user2);
+        jetStaking.unstakeAll();
+        vm.stopPrank();
+        vm.warp(block.timestamp + 2 days + 1);
+        vm.prank(user1);
+        jetStaking.withdrawAll();
+        vm.prank(user2);
+        jetStaking.withdrawAll();
+        assert(aurora.balanceOf(user1) == aurora.balanceOf(user2));
+    }
+
+    /**
+     * @dev Stakes 10**19 aurora tokens for user2, waits 10 weeks, then stakes MUCH
+     *      more aurora on behalf of user3 (10**30). Then unstakes for user1. This 
+     *      test ensures that a "whale sniper" cannot affect the rewards of other 
+     *      people 
+     */
+    function testWhaleSniper() public { 
+        stakeFor(user2, 10 ether);
+        vm.warp(block.timestamp + 10 weeks);
+        stakeFor(user3, 10**30); // No matter how much is staked rewards arent diluted
+        vm.startPrank(user2);
+        jetStaking.moveAllRewardsToPending();
+        jetStaking.unstakeAll();
+        vm.warp(block.timestamp + 2 days + 1);
+        jetStaking.withdrawAll();
+        assert(aurora.balanceOf(user2) > 10 ether);
+    }    
+
+     /**
+     * @dev Stakes 100 aurora on behalf of address(this). Then stakes 1 aurora on behalf 
+     *      of user1. Then immediately unstakes and checks that the balance of user1 is
+     *      greater than 1 (should technically be 1). This is because as the share price 
+     *      goes up, a user can pocket the difference between the share price and 1 aurora.
+     *      However, given we are speaking about 0.(17 zeroes later)1 aurora tokens, the 
+     *      value is hugely outweighed by gas cost.  
+     */
+    function testStake1Aurora() public {
+        uint256 amount = 100 ether;
+        aurora.approve(address(jetStaking), amount);
+        jetStaking.stake(amount);
+        vm.warp(block.timestamp + 10 weeks);
+        stakeFor(user1, 1);
+        vm.startPrank(user1);
+        jetStaking.unstakeAll();
+        vm.warp(block.timestamp + 2 days + 1);
+        jetStaking.withdrawAll();
+        assert(aurora.balanceOf(user1) > 1);
```

Comment:
Was user1 balance before staking > 1 ?, if so, this assertion should consider the old balance + the newly generated aurora rewards.